File: 0149.MaxPointsonaLine.py
# Definition for a point.
class Point(object):
    def __init__(self, a=0, b=0):
        self.x = a
        self.y = b

class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """
        length = len(points)
        if length <= 2:
            return length
        ans = 0
        for i in xrange(length-1):
            dic = {'i': 1}
            same = 0
            xi, yi = points[i].x, points[i].y
            for j in xrange(i+1, length):
                xj, yj = points[j].x, points[j].y
                if xi == xj and yi == yj:
                    same += 1
                    continue
                if xi == xj:
                    scope = 'i'
                else:
                    scope = float(yi-yj)/(xi-xj)
                if scope not in dic:
                    dic[scope] = 1
                dic[scope] += 1
            ans = max(ans, max(dic.values())+same)
        return ans

print Solution().maxPoints([Point(0,0), Point(1,1), Point(2,2), Point(1,-1)])

File: 0149.MaxPointsonaLine.go
func maxPoints(points [][]int) int {
    n := len(points)
    if n <= 2 {
        return n
    }
    ans := 0
    for i := 0; i < n; i++ {
        Map := make(map[[2]int]int)
        same, curMax := 0, 0
        for j := i+1; j < n; j++ {
            x, y := points[i][0]-points[j][0], points[i][1]-points[j][1]
            if x == 0 && y == 0 {
                same++
                continue
            }
            g := gcd(x, y)
            if g != 0 {
                x, y = x/g, y/g
            }
            Map[[2]int{x,y}]++
            curMax = max(curMax, Map[[2]int{x,y}])
        }
        ans = max(ans, curMax+same+1)
    }
    return ans
}

func gcd(x, y int) int {
    for y != 0 {
        x, y = y, x%y
    }
    return x
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}